tqdm/contrib/sbar.py

```diff
@@ -0,0 +1,247 @@
+    sbar is a subclass of `tqdm.autonotebook.tqdm` to make using non-iterables
+    steps easier. 
+
+    It takes a list of "steps". 
+    
+    "Wait, didn't you say it was for non-iterables?"
+    Yup.
+    For example, suppose you have a pytorch_lightning.LightningDataModule
+    ```python
+    class MyDM(pl.LightningDataModule):
+        _STEPS = [...]
+        def __init__(self, *args, **kwargs):
+            # ...
+            self.sbar(self._STEPS, desc='MyDM')
+
+        ...
+        
+    ```
+
+    and you call the `prepare` method which downloads and transforms data.
+    You can easily see where you might have several non-standard methods
+    that are part of a pipeline, which might time some time to resolve. 
+    You can now have a `sbar` instance living inside you class and more easily
+    move between "steps" of the preprocessing pipeline (method calls)
+
+    e.g.
+
+    ```python
+    _STEPS = [     
+        ProgressStep(1, 20, desc='download'),
+
+        ProgressStep(2, 5, 0, desc='load_timepoints'),
+        ProgressStep(2, 5, 1, desc='load_timepoints'),
+        ProgressStep(2, 5, 2, desc='load_timepoints'),
+        ProgressStep(2, 5, 3, desc='load_timepoints'),
+        ProgressStep(2, 5, 4, desc='load_timepoints'),
+        # ProgressStep(2, val=25, desc='load_timepoints'), # <-- doesn't need to be explicitly stated
+
+        ProgressStep(3, 5, 0, desc='data_wrangling'),
+        ProgressStep(3, 5, 1, desc='data_wrangling'),
+        ProgressStep(3, 5, 2, desc='data_wrangling'),
+        ProgressStep(3, 5, 3, desc='data_wrangling'),
+        ProgressStep(3, 5, 4, desc='data_wrangling'),
+        # ProgressStep(3, val=25, desc='data_wrangling'),
+
+        ProgressStep(4, 5, desc='merge_data'),
+
+        ProgressStep(5, 5, 0, desc='filter_data'),
+        ProgressStep(5, 5, 1, desc='filter_data'),
+        ProgressStep(5, 5, 2, desc='filter_data'),
+        ProgressStep(5, 5, 3, desc='filter_data'),
+        ProgressStep(5, 5, 4, desc='filter_data'),
+        ProgressStep(5, 5, 5, desc='filter_data'),
+        # ProgressStep(5, val=30, desc='filter_data'),
+
+        ProgressStep(6, 20, desc='embed'),
+
+        ProgressStep(7, desc='save'),
+    ]
+
+    ```
+    
+    which can be used inside the `MyDM` class like:
+
+    ```python
+    # inside MyDM class delcaration
+    def prepare(self):
+        # ...
+        self.sbar.write('Downloading data')
+        self.sbar.step(1)
+        
+        # ...
+        self.sbar.write('Loading datafiles')
+        for i in range(5):
+            self.sbar.step(2, i)    # <--- handles substeps
+        
+        # ...
+
+        # ...
+        self.sbar.step(7)                
+    ```
+
+    Yes the `_STEPS` list is an iterable and technically compatible with
+    `tqdm` _as is_. However, handling abstract method calling sounds like 
+    a larger pain than this.
+    
+    
+    '''
+    __NUM_IDX = 0    
+    __SUB_IDX = 2
```

Comment:
Since right now each `step` is in essence a tuple, these internal class  properties makes it easy to readily switch the ordering in case `main_step_num, step_val, sub_step_num` isn't the desired order